packages/@jsii/kernel/src/disk-cache/disk-cache.ts

```diff
@@ -0,0 +1,244 @@
+      throw new Error(`Cache entry key must contain at least 1 element!`);
+    }
+
+    return new Entry(
+      join(
+        this.#root,
+        ...key.flatMap((s) =>
+          s
+            .replace(/[^@a-z0-9_.\\/-]+/g, '_')
+            .split(/[\\/]+/)
+            .map((ss) => {
+              if (ss === '..') {
+                throw new Error(
+                  `A cache entry key cannot contain a '..' path segment! (${s})`,
+                );
+              }
+              return ss;
+            }),
+        ),
+      ),
+    );
+  }
+
+  public entryFor(path: string, ...comments: readonly string[]): Entry {
+    const rawDigest = digestFile(path, ...comments);
+    return this.entry(...comments, rawDigest.toString('hex'));
+  }
+
+  public pruneExpiredEntries() {
+    const cutOff = new Date(Date.now() - PRUNE_AFTER_MILLISECONDS);
+    for (const entry of this.entries()) {
+      if (entry.atime < cutOff) {
+        entry.lock((lockedEntry) => {
+          // Check again in case it's been accessed which we waited for the lock...
+          if (entry.atime > cutOff) {
+            return;
+          }
+          lockedEntry.delete();
+        });
+      }
+    }
+
+    for (const dir of directoriesUnder(this.#root, true)) {
+      if (process.platform === 'darwin') {
+        try {
+          rmSync(join(dir, '.DS_Store'), { force: true });
+        } catch {
+          // Ignore errors...
+        }
+      }
+      if (readdirSync(dir).length === 0) {
+        try {
+          rmdirSync(dir);
+        } catch {
+          // Ignore errors, directory may no longer be empty...
+        }
+      }
+    }
+  }
+
+  private *entries(): Generator<Entry, void, void> {
+    yield* inDirectory(this.#root);
+
+    function* inDirectory(dir: string): Generator<Entry, void, void> {
+      if (existsSync(join(dir, MARKER_FILE_NAME))) {
+        return yield new Entry(dir);
+      }
+      for (const file of directoriesUnder(dir)) {
+        yield* inDirectory(file);
+      }
+    }
+  }
+}
+
+export class Entry {
+  public constructor(public readonly path: string) {}
+
+  public get atime(): Date {
+    try {
+      const stat = statSync(this.markerFile);
+      return stat.atime;
+    } catch (err: any) {
+      if (err.code !== 'ENOENT') {
+        throw err;
+      }
+      return new Date(0);
+    }
+  }
+
+  public get pathExists() {
+    return existsSync(this.path);
+  }
+
+  private get lockFile(): string {
+    return `${this.path}.lock`;
+  }
+
+  private get markerFile(): string {
+    return join(this.path, MARKER_FILE_NAME);
+  }
+
+  public lock<T>(cb: (entry: LockedEntry) => T): T {
+    mkdirSync(dirname(this.path), { recursive: true });
+    lockSync(this.lockFile, { retries: 12, stale: 5_000 });
+    let disposed = false;
+    try {
+      return cb({
+        delete: () => {
+          if (disposed) {
+            throw new Error(
+              `Cannot delete ${this.path} once the lock block as returned!`,
```

Comment:
[nitpick] Is `as` supposed to be `has`, or should a verb be part of the second sentence?